tortoisebeacon/tortoise_beacon.go

```diff
@@ -193,76 +203,178 @@ func (tb *TortoiseBeacon) Close() {
+		tb.beaconsFromBlocks[epochID] = make(map[string]*epochBeacon)
+	}
+	key := string(beacon)
+	entry, ok := tb.beaconsFromBlocks[epochID][key]
+	if !ok {
+		tb.beaconsFromBlocks[epochID][key] = &epochBeacon{
+			weight: weight,
+			blocks: map[types.BlockID]struct{}{blockID: {}},
 		}
+		tb.logger.With().Debug("added beacon from block",
+			epochID,
+			blockID,
+			log.String("beacon", types.BytesToHash(beacon).ShortString()),
+			log.Uint64("weight", weight))
+		return
 	}
 
-	if (epochID - 1).IsGenesis() {
-		return types.HexToHash32(genesisBeacon).Bytes(), nil
+	// checks if we have recorded this blockID before
+	if _, ok := entry.blocks[blockID]; ok {
+		tb.logger.With().Warning("block already reported beacon", epochID, blockID)
+		return
 	}
 
+	entry.blocks[blockID] = struct{}{}
+	entry.weight += weight
+	tb.logger.With().Debug("added beacon from block",
+		epochID,
+		blockID,
+		log.String("beacon", types.BytesToHash(beacon).ShortString()),
+		log.Uint64("weight", weight))
+}
+
+func (tb *TortoiseBeacon) findMostWeightedBeaconForEpoch(epoch types.EpochID) []byte {
 	tb.mu.RLock()
 	defer tb.mu.RUnlock()
-
-	beacon, ok := tb.beacons[epochID-1]
+	epochBeacons, ok := tb.beaconsFromBlocks[epoch]
 	if !ok {
-		tb.logger.With().Error("beacon is not calculated",
-			log.Uint32("target_epoch", uint32(epochID)),
-			log.Uint32("beacon_epoch", uint32(epochID-1)))
+		return nil
+	}
+	var mostWeight uint64
+	var beacon []byte
+	numBlocks := 0
+	for k, v := range epochBeacons {
+		if v.weight > mostWeight {
+			beacon = []byte(k)
+			mostWeight = v.weight
+		}
+		numBlocks += len(v.blocks)
+	}
+
+	logger := tb.logger.WithFields(epoch, log.Int("num_blocks", numBlocks))
+
+	if uint32(numBlocks) < tb.config.BeaconSyncNumBlocks {
+		logger.Debug("not enough blocks to determine beacon")
+		return nil
+	}
+
+	logger.With().Info("beacon determined for epoch",
+		log.String("beacon", types.BytesToHash(beacon).ShortString()),
+		log.Uint64("weight", mostWeight))
+	return beacon
+}
+
+// GetBeacon returns a Tortoise Beacon value as []byte for a certain epoch or an error if it doesn't exist.
+// TODO(nkryuchkov): consider not using (using DB instead)
+func (tb *TortoiseBeacon) GetBeacon(epochID types.EpochID) ([]byte, error) {
+	if epochID == 0 {
+		return nil, ErrZeroEpoch
+	}
 
-		return nil, ErrBeaconNotCalculated
+	beaconEpoch := epochID - 1
+	if beaconEpoch.IsGenesis() {
+		return types.HexToHash32(genesisBeacon).Bytes(), nil
 	}
 
-	return beacon.Bytes(), nil
+	beacon := tb.getBeacon(beaconEpoch)
+	if beacon != nil {
+		return beacon, nil
+	}
+
+	beacon = tb.getPersistedBeacon(beaconEpoch)
+	if beacon != nil {
+		return beacon, nil
+	}
+
+	tb.logger.With().Warning("beacon not available",
+		epochID,
+		log.Uint32("beacon_epoch", uint32(beaconEpoch)))
+	return nil, ErrBeaconNotCalculated
+}
+
+func (tb *TortoiseBeacon) getBeacon(epoch types.EpochID) []byte {
+	tb.mu.RLock()
+	defer tb.mu.RUnlock()
+	if beacon, ok := tb.beacons[epoch]; ok {
```

Comment:
why not to store data only in db? it is not going to be a bottleneck

Reply:
it is not going to be a bottleneck for sure but i still don't like to access disk every time a beacon is queried.

our protocol is based on synchronous network. and some messages are time sensitive. the hardware assumption for home miners are **not** SSD but HDD. i think we should still avoid disk seeks and the long tail on disk read time as much as we can. the epoch numbers to keep is limited to 10. i'm not concerned abt memory usage.